### PR TITLE
man: Advertise systemd-time-wait-sync.service more

### PR DESCRIPTION
**Pull request by Minoru**
The description of time-sync.target says that NTP services *should* pull
that target, but doesn't mention that e.g. systemd-timesyncd.service
doesn't actually do that. As a result, time-sync.target is reached way
earlier than people expect; see #5097, #8861, #11008.

systemd provides systemd-time-wait-sync.service to ameliorate this
problem, but doesn't feature it prominently in relevant manpages. In
fact, it's only mentioned in passing in systemd-timesyncd.service(8). As
a result, I ended up re-implementing that service, and I'm not the first
one: https://github.com/NixOS/nixpkgs/pull/51338

This patch adds a mention right in the description of time-sync.target,
which will hopefully raise awareness of this helper service.

### CODE REVIEW

File: man/systemd.special.xml [hunk 1]
```diff
@@ -1018,7 +1018,14 @@
...
             <varname>After=</varname> for this target unit to all SysV
             init script service units with an LSB header referring to
-            the <literal>$time</literal> facility. </para>
+            the <literal>$time</literal> facility, and also to all timer
+            units with at least one <varname>OnCalendar=</varname>
+            directive. </para>
+
+            <para>Unless
+            <citerefentry><refentrytitle>systemd-time-wait-sync.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>
+            is enabled, this target might get reached way before the
```
Comment (yuwata) [inline man/systemd.special.xml]: 'get reached way'? Maybe, 'get reached' is enough. No?

**Comment by yuwata** [comment]
@poettering Why time-sync.target is not an active unit? The unit looks something like network-online.target.

**Comment by keszybz** [comment]
I think this needs a bigger update. The initial problem is that the existing description of `time-sync.target` is misleading. Maybe something like this:
```
A target that gathers services responsible for synchronizing the system clock from a remote source.
Services that implement such synchronization (NTP clients and similar) should pull in this target and order themselves before it.
All services where correct time is essential should be ordered after this unit, but not pull it in.
<command>systemd</command> automatically adds dependencies of type <varname>After=</varname> for this target unit to all SysV init script service units with an LSB header referring to the <literal>$time</literal> facility.

This target has two different meanings:
- when systemd-time-wait-sync.service is disabled (the default), it only means that synchronization services have been started.
- when systemd-time-wait-sync.service is enabled, this target will be delayed until synchronization is achieved.
Use `systemctl enable systemd-time-wait-sync.service` if services ordered after this target require precise time at all times.

Hint: if only approximate time is required, <filename>time-set.target</filename> might be more appropriate, see above.
```

**Comment by keszybz** [comment]
An alternative would be to just enable `systemd-time-wait-sync.service` be default. I expect that the majority of people expect that `time-sync.target` does what the name says, i.e. guarantees time synchronization. I know we didn't do that on purpose, but the repeated bugs opened by people who are surprised by current semantics suggest that *status quo* is unintuitive.

File: man/systemd.special.xml (hunk 1 above)
Comment (Minoru) [reply man/systemd.special.xml]: Agreed, the language is too informal for a manpage. I force-pushed an update that removes "way" and changes "time" to "clock".

**Comment by Minoru** [comment]
@keszybz I gather that `systemd-time-wait-sync.service` can't be enabled by default because there is no guarantee that the user is running `systemd-timesyncd.service`. For example, on Armbian this service is enabled, but is conditioned on the absence of other NTP implementations; however, Armbian ships Chrony by default. As a result, if `systemd-time-wait-sync.service` were to be enabled by default on Armbian, I'd get occasional lock-ups during boot (https://github.com/systemd/systemd/issues/8683).

I like your proposed wording, but I'm not sure such a level of detail is appropriate for a man page. I always like man pages are references, not ready-made tutorials; they merely present the facts without any guidance on how to actually apply them (except maybe in a couple paragraphs in the EXAMPLES section).

It's also my first time contributing to systemd, so pardon a silly question: who's the ultimate decision-maker for this PR? Should I wait for Poettering to state his opinion, and treat your comments as suggestions, or do I work with you and @yuwata?

Finally: thanks for taking a look at the changes!

**Comment by Minoru** [comment]
Bump. Is there anything I can do to get this merged?

**Comment by bluca** [comment]
> Bump. Is there anything I can do to get this merged?

Sorry for the delay - but there is an ongoing discussion in https://github.com/systemd/systemd/pull/17742 related to this, so we should wait for a resolution there first I think.

File: man/systemd.special.xml [hunk 2]
```diff
@@ -1018,7 +1018,14 @@
             in. systemd automatically adds dependencies of type
             <varname>After=</varname> for this target unit to all SysV
             init script service units with an LSB header referring to
-            the <literal>$time</literal> facility. </para>
+            the <literal>$time</literal> facility, and also to all timer
+            units with at least one <varname>OnCalendar=</varname>
+            directive. </para>
+
+            <para>Unless
+            <citerefentry><refentrytitle>systemd-time-wait-sync.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>
```
Comment (poettering) [inline man/systemd.special.xml:1027]: maybe also say "(or an equivalent service for alternative NTP implementations)"

Thing is: other NTP implementations might provide their own implementation of such a service, and it's entirely OK and sufficient to enable those, without enabling systemd-time-wait-sync.service

File: man/systemd.special.xml [hunk 3]
```diff
@@ -1018,7 +1018,14 @@
...
+            is enabled, this target might get reached before the clock
+            is actually synchronized.</para>
```
Comment (poettering) [inline man/systemd.special.xml]: say "…synchronized to any accurate reference clock" or so. Just saying "synchronized" doesn't mean much, it is after all syncrhonized to the systems local RTC...

**Comment by poettering** [comment]
> An alternative would be to just enable `systemd-time-wait-sync.service` be default. I expect that the majority of people expect that `time-sync.target` does what the name says, i.e. guarantees time synchronization. I know we didn't do that on purpose, but the repeated bugs opened by people who are surprised by current semantics suggest that _status quo_ is unintuitive.

I don#t think this should be enabled by default, the s-t-w-s.s service delays the boot process, depending on external conditions. That's something that has to be strictly opt-in I am sure. I certainly don't want my laptop to wait on boot for a NTP sync. 

Now, of course only the parts of the boot that are odered after time-sync.target will actually be delayed if s-t-w-s.s is enabled, but that's all .timer units with OnCalendar=, i.e. already a lot. 

Also, the way I see it s-t-w-s.s is one implementation of the concept, but other NTP services might have their own implementations that using some NTP server specific IPC to implement what they need to implement. We shouldn't enable our implementation hence, because it really depends on the NTP server if it's actually the right one to use.

**Comment by poettering** [comment]
> This target has two different meanings:
> - when systemd-time-wait-sync.service is disabled (the default), it only means that synchronization services have been started.

It doesn't really mean that. If you have no NTP server enabled it doesn't mean that...

if s-t-w-s.s (or some equivalent service) is not enabled it just means the same as time-set.target.

And time-set.target means that that if some time serveice is enabled that can make time monotonic via some local touch file that is has completed. (e.g. systemd-timesyncd.service is one of those). If such as service is not enabled either, then both targets means exactly nothing.

File: man/systemd.special.xml (hunk 2 above)
Comment (Minoru) [reply man/systemd.special.xml:1027]: Applied, thanks! The sentence got a bit too long to my taste, so I split it in two.

**Comment by bluca** [comment]
The two comments have been applied pretty much verbatim, and the manpage with this is already much clearer than before, hence merging. If we want further updates we can always follow up later.

But for next time, please rebase and squash the fixup immediately.

Thanks!

**Pull request merged**